### PR TITLE
4. domaća zadaća

### PR DESCRIPTION
**Pull request by danijelLoc**

### CODE REVIEW

File: QuizApp/MVP/Model/Repository/QuizDatabaseDataSource.swift [hunk 1]
```diff
@@ -15,4 +15,87 @@ class QuizDatabaseDataSource{
...
+
+            do {
+                try coreDataContext.save()
+            } catch {
+                print("#### Error when saving updated quiz: \(error)")
+            }
+        }
+    }
+    
+    private func fetcQuiz(withId id: Int) throws -> CDQuiz? {
```
Comment (dorastipkovic) [inline QuizApp/MVP/Model/Repository/QuizDatabaseDataSource.swift]: Typo in the function name.

File: QuizApp/MVP/Model/Repository/QuizRepository.swift [hunk 2]
```diff
@@ -8,13 +8,98 @@
...
+    public let quizNetworkSource:QuizNetworkDataSource!
+    private var imagesDictionary: [Int:Data?] = [:]
     
     init(quizDatabaseSource:QuizDatabaseDataSource, quizNetworkSource:QuizNetworkDataSource) {
         self.quizDatabaseSourece = quizDatabaseSource
         self.quizNetworkSource = quizNetworkSource
     }
+    
+    public func getQuizzes(presenter:QuizzesPresenter){
+        // print("Get quizzes")
```
Comment (dorastipkovic) [inline QuizApp/MVP/Model/Repository/QuizRepository.swift]: Remove unnecessary comments, they only make code less readable.

File: QuizApp/MVP/Model/Repository/QuizRepository.swift [hunk 3]
```diff
@@ -8,13 +8,98 @@
...
+                    presenter.delegate.showNoQuizzes()
+                }else{
+                    presenter.proccesAndShowQuizzes(allQuizzes: quizzes)
+                }
+            }
+        }
+    }
+    
+    
+    func showError(error:RequestError,presenter:QuizzesPresenter){
```
Comment (dorastipkovic) [inline QuizApp/MVP/Model/Repository/QuizRepository.swift]: This does not belong in QuizRepository. Your QuizzesPresenter has a reference to this class, so ideally you don't want to have a two way reference.

File: QuizApp/MVP/Presenter/LoginPresenter.swift [hunk 4]
```diff
@@ -45,7 +45,13 @@ class LoginPresenter{
             defaults.set(loginResponse.userId, forKey: "user_id")
             defaults.set(loginResponse.token, forKey: "user_token")
             print("username:\(username), password:\(password),\nuserId:\(loginResponse.userId), token:\(loginResponse.token)")
-            self.delegate.showGoodLogin()
+            self.presentGoodLogin()
+        }
+    }
+    
+    func presentGoodLogin(){
```
Comment (dorastipkovic) [inline QuizApp/MVP/Presenter/LoginPresenter.swift]: Successful would make more sense than Good.

File: QuizApp/MVP/ViewController/LoginViewController.swift [hunk 5]
```diff
@@ -17,19 +17,25 @@ class LoginViewController : UIViewController, LoginViewDelegate {
...
+        // remember login
+        let defaults = UserDefaults.standard
+        guard let _ = defaults.object(forKey: "user_id"),
+              let _ = defaults.object(forKey: "user_token")
+        else{
+            buildViews()
+            return
+        }
+        self.styleViews()
+        self.presenter.presentGoodLogin()
```
Comment (dorastipkovic) [inline QuizApp/MVP/ViewController/LoginViewController.swift]: Why would you call this here? Isn’t this skipping the whole login process?

File: QuizApp/MVP/ViewController/QuestionViewController.swift [hunk 6]
```diff
@@ -21,10 +20,9 @@ class QuestionViewController : UIViewController,QuestionViewDelegate {
     
     convenience init(router: AppRouterProtocol, question: Question, qvc: QuizViewController) {
         self.init()
-        self.router = router
         self.question = question
         self.qvc = qvc
-        self.presenter = QuestionPresenter(delegate: self)
+        self.presenter = QuestionPresenter(delegate: self, router: router)
```
Comment (dorastipkovic) [inline QuizApp/MVP/ViewController/QuestionViewController.swift]: You should initialize this presenter in the Router, not inside UIViewController

File: QuizApp/MVP/ViewController/QuizResultViewController.swift [hunk 7]
```diff
@@ -24,9 +21,8 @@ class QuizResultViewController : UIViewController,QuizResultViewDelegate {
     
     convenience init(router: AppRouterProtocol, result: QuizResult) {
         self.init()
-        self.router = router
         self.quizResult = result
-        self.presenter = QuizResultPresenter(delegate: self, quizResult: quizResult)
+        self.presenter = QuizResultPresenter(router: router, delegate: self, quizResult: quizResult)
```
Comment (dorastipkovic) [inline QuizApp/MVP/ViewController/QuizResultViewController.swift]: Again, presenters should be intialized inside Router.

**Review by dorastipkovic — CHANGES_REQUESTED** [review]

File: QuizApp/MVP/ViewController/LoginViewController.swift (hunk 5 above)
Comment (danijelLoc) [reply QuizApp/MVP/ViewController/LoginViewController.swift]: If user is logged in, aplication remembers it and skips login

**Pull request merged**